### PR TITLE
Add ShieldGemma2 support with Gemma3 dependency

### PR DESCRIPTION
**Pull request by Sainalu**
## PR 类型
新增模型支持

## PR 说明
新增 `ShieldGemma2` 在 PaddleFormers 中的支持。

由于 `ShieldGemma2` 依赖 `Gemma3` 主干实现，因此本PR一并携带 `Gemma3` 依赖代码。对应 `gemma3` 支持也已拆分为独立 PR。

本 PR 包含：

- 在 `paddleformers/transformers/shieldgemma2` 下新增模型组网代码
- 在 `paddleformers/transformers/gemma3` 下携带 `ShieldGemma2` 所需依赖实现
- 补充 `AutoConfig` / `AutoModel` 相关注册
- 新增 `tests/transformers/shieldgemma2/test_modeling.py`
- 完成 Torch/HF 与 Paddle 的推理和训练对齐验证

## 推理对齐情况
对同一输入样本进行 Torch/HF 与 Paddle 前向对齐，结果如下：

- `top-10 token ids` 完全一致
- `yes/no` 分类 logits 对齐到 `1e-5` 量级

对应 `top-10 token ids`：
- Paddle: `[103085, 14553, 121486, 157028, 238170, 47609, 36008, 217152, 105916, 73655]`
- Torch/HF: `[103085, 14553, 121486, 157028, 238170, 47609, 36008, 217152, 105916, 73655]`

## ms-swift注册
  1. 在 ms-swift/swift/model/constant.py 里加一个模型类型

    class MLLMModelType:
        ...
        gemma3_vision = 'gemma3_vision'
        shieldgemma2 = 'shieldgemma2'

  2. 在 ms-swift/swift/model/models/gemma.py 里新增注册
模板复用 gemma3_vision

    import os
    from types import MethodType
  
    from transformers import GenerationConfig, PreTrainedModel
  
    from swift.template import TemplateType
    from ..constant import LLMModelType, MLLMModelType
    from ..model_arch import ModelArch
    from ..model_meta import Model, ModelGroup, ModelMeta
    from ..patcher import patch_output_to_input_device
    from ..register import ModelLoader, SentenceTransformersLoader, register_model
  
  
    class ShieldGemma2Loader(ModelLoader):
  
        def get_config(self, model_dir):
            # keep aligned with Gemma3 training path
            self.attn_impl = self.attn_impl or 'eager'
            return super().get_config(model_dir)
  
        def get_model(self, model_dir: str, config, processor, model_kwargs) -> PreTrainedModel:
            from torch import nn
            from transformers import ShieldGemma2ForImageClassification
  
            model = ShieldGemma2ForImageClassification.from_pretrained(
                model_dir,
                config=config,
                trust_remote_code=True,
                **model_kwargs,
            )
  
            # ms-swift sft path expects generation_config on the outer model
            if not hasattr(model, 'generation_config') or model.generation_config is None:
                inner_generation_config = getattr(model.model, 'generation_config', None)
                model.generation_config = inner_generation_config or GenerationConfig.from_model_config(model.config)
  
            # optional: keep validation deterministic
            if os.environ.get('DISABLE_DROPOUT', '1') == '1':
                for module in model.modules():
                    if isinstance(module, nn.Dropout):
                        module.p = 0.0
                    for attr in ('dropout', 'attention_dropout', 'activation_dropout'):
                        if hasattr(module, attr):
                            value = getattr(module, attr)
                            if isinstance(value, (int, float)):
                                setattr(module, attr, 0.0)
  
            original_forward = model.forward
  
            def _forward_with_lm_loss(self, *args, **kwargs):
                # training loss should come from inner Gemma3 conditional generation model
                if kwargs.get('labels', None) is None:
                    return original_forward(*args, **kwargs)
  
                inner_kwargs = {
                    'input_ids': kwargs.get('input_ids'),
                    'attention_mask': kwargs.get('attention_mask'),
                    'position_ids': kwargs.get('position_ids'),
                    'token_type_ids': kwargs.get('token_type_ids'),
                    'pixel_values': kwargs.get('pixel_values'),
                    'inputs_embeds': kwargs.get('inputs_embeds'),
                    'labels': kwargs.get('labels'),
                    'use_cache': False,
                    'return_dict': kwargs.get('return_dict', True),
                }
                inner_kwargs = {k: v for k, v in inner_kwargs.items() if v is not None}
                return self.model(**inner_kwargs)
  
            model.forward = MethodType(_forward_with_lm_loss, model)
            return model
  
  
    register_model(
        ModelMeta(
            MLLMModelType.shieldgemma2,
            [
                ModelGroup([
                    Model('google/shieldgemma-2-4b-it', 'google/shieldgemma-2-4b-it'),
                ]),
            ],
            ShieldGemma2Loader,
            template=TemplateType.gemma3_vision,
            architectures=['ShieldGemma2ForImageClassification'],
            model_arch=ModelArch.llava_hf,
            requires=['transformers>=4.49'],
            task_type='causal_lm',
            tags=['vision'],
        ))
 3. 训练时直接复用 gemma3_vision
 
## 训练对齐情况
使用 GSM8K 数据进行训练对齐验证（）。

| paddle loss | torch loss | diff |
| --- | --- | --- |
| 16.00169373 | 16.49606705 | -0.49437332 |
| 14.71333885 | 16.95349312 | -2.24015427 |
| 12.22286510 | 19.18050003 | -6.95763493 |
| 11.21158791 | 17.87071991 | -6.65913200 |
| 9.66808033 | 18.10136414 | -8.43328381 |
| 8.44376850 | 19.01485062 | -10.57108212 |
| 7.67826653 | 16.51401329 | -8.83574676 |
| 9.85611057 | 18.44094467 | -8.58483410 |
| 8.03356457 | 17.96991920 | -9.93635463 |
| 7.60740328 | 13.71025372 | -6.10285044 |
| 8.77243137 | 11.98634434 | -3.21391297 |
| 5.47642279 | 10.85439682 | -5.37797403 |
| 5.80109024 | 10.15167332 | -4.35058308 |
| 4.89205027 | 9.51164722 | -4.61959695 |
| 4.98916817 | 8.14846134 | -3.15929317 |
| 4.12202406 | 8.63837433 | -4.51635027 |
| 4.26863337 | 5.97669315 | -1.70805978 |
| 7.10579062 | 6.73086739 | 0.37492323 |
| 5.35105228 | 8.19053936 | -2.83948708 |
| 3.56379008 | 5.55994272 | -1.99615264 |
| 3.37518883 | 5.08673048 | -1.71154165 |
| 6.67938042 | 4.69946909 | 1.97991133 |
| 6.33457422 | 5.04648590 | 1.28808832 |
| 3.86823177 | 6.01899004 | -2.15075827 |
| 3.61999869 | 3.50136161 | 0.11863708 |
| 5.57703447 | 5.07118320 | 0.50585127 |
| 5.52485657 | 6.53269911 | -1.00784254 |
| 3.90741539 | 5.76341677 | -1.85600138 |
| 5.30978870 | 5.22436953 | 0.08541917 |
| 6.61713266 | 3.10578084 | 3.51135182 |
| 3.20617509 | 4.87641954 | -1.67024445 |
| 2.99516487 | 4.48868513 | -1.49352026 |
| 5.53285885 | 2.83593726 | 2.69692159 |
| 4.78122711 | 5.17554760 | -0.39432049 |
| 2.10321355 | 3.65890551 | -1.55569196 |
| 4.88998461 | 2.08641648 | 2.80356813 |
| 2.06521869 | 3.32727408 | -1.26205539 |
| 1.92053914 | 2.85104012 | -0.93050098 |
| 3.36463404 | 2.82646298 | 0.53817106 |
| 4.01963854 | 3.67929554 | 0.34034300 |
| 2.65473366 | 5.33922195 | -2.68448829 |
| 2.35188341 | 3.20224285 | -0.85035944 |
| 2.24985266 | 4.37660599 | -2.12675333 |
| 2.99977708 | 2.32494974 | 0.67482734 |
| 2.20928407 | 3.00671077 | -0.79742670 |
| 2.32395935 | 4.41969490 | -2.09573555 |
| 4.42433691 | 3.53861713 | 0.88571978 |
| 2.21135378 | 3.84755135 | -1.63619757 |
| 4.37625074 | 2.29686213 | 2.07938861 |
| 2.24772120 | 2.04504347 | 0.20267773 |
| 2.45447421 | 1.83361006 | 0.62086415 |
| 4.53304338 | 3.72673273 | 0.80631065 |
| 2.49710369 | 2.10875511 | 0.38834858 |
| 3.48686075 | 2.74079490 | 0.74606585 |
| 3.55673051 | 2.15397930 | 1.40275121 |
| 2.35233259 | 3.11158371 | -0.75925112 |
| 2.82058597 | 3.97401214 | -1.15342617 |
| 3.73818088 | 2.48388839 | 1.25429249 |
| 2.83329129 | 1.84142590 | 0.99186539 |
| 3.82988954 | 2.58507085 | 1.24481869 |
| 4.09733868 | 3.51621914 | 0.58111954 |
| 2.09185266 | 4.36878872 | -2.27693606 |
| 1.99270928 | 3.77091718 | -1.77820790 |
| 3.53731775 | 2.86979246 | 0.66752529 |
| 2.35056400 | 3.31573987 | -0.96517587 |
| 1.31469834 | 3.76502347 | -2.45032513 |
| 2.56759191 | 3.31528974 | -0.74769783 |
| 3.16971564 | 2.83173370 | 0.33798194 |
| 2.70302248 | 2.19158316 | 0.51143932 |
| 3.16695642 | 2.82044482 | 0.34651160 |
| 1.68660164 | 3.03851414 | -1.35191250 |
| 3.12188959 | 1.33334994 | 1.78853965 |
| 4.99617386 | 2.54442596 | 2.45174790 |
| 3.52140188 | 3.23061705 | 0.29078483 |
| 2.20802951 | 4.11522722 | -1.90719771 |
| 1.53144503 | 1.60658860 | -0.07514357 |
| 2.28293943 | 2.51524425 | -0.23230482 |
| 1.99344146 | 1.46567798 | 0.52776348 |
| 1.85814416 | 3.14762330 | -1.28947914 |
| 2.49384975 | 2.33078742 | 0.16306233 |
| 3.00237036 | 2.72561526 | 0.27675510 |
| 2.30961776 | 1.85101843 | 0.45859933 |
| 2.36047268 | 3.57117987 | -1.21070719 |
| 2.83624601 | 3.72562003 | -0.88937402 |
| 3.23331094 | 2.25343466 | 0.97987628 |
| 5.01312160 | 4.18181515 | 0.83130645 |
| 3.19645905 | 2.43944168 | 0.75701737 |
| 1.45707142 | 2.23586464 | -0.77879322 |
| 4.30001736 | 2.67853880 | 1.62147856 |
| 3.68523908 | 2.71893525 | 0.96630383 |
| 2.99299622 | 1.03801286 | 1.95498336 |
| 4.10749722 | 3.35801315 | 0.74948407 |
| 2.27043128 | 3.22624874 | -0.95581746 |
| 0.90721697 | 2.50921869 | -1.60200172 |
| 1.42224860 | 2.37878108 | -0.95653248 |
| 2.78236747 | 3.61671495 | -0.83434748 |
| 1.72835886 | 1.27770722 | 0.45065164 |
| 4.10920906 | 2.97133064 | 1.13787842 |

说明：
- 当前模型前向已对齐
- 训练曲线对齐
- 本 PR 携带 `gemma3` 依赖实现，因此会与 `gemma3` PR 存在重叠文件
- 该模型为多模态判别模型，实际使用训练应该采用其他数据集进行，本训练对齐仅为验证

### CODE REVIEW

**Comment by a31413510** [comment]
需要确保CI跑通，如果不是自己的问题，可以评论 /re-run all-failed 重跑

**Comment by a31413510** [comment]
ci存在问题，本地代码需要同步到最新，再重新提交

**Comment by a31413510** [comment]
需要解决代码冲突问题

**Comment by a31413510** [comment]
本地代码更新到最新，解决该PR的代码冲突
image

File: paddleformers/transformers/auto/configuration.py [hunk 1]
```diff
@@ -107,6 +109,8 @@
         ("glm_ocr", "GlmOcrForConditionalGeneration"),
         ("minicpm", "MiniCPM"),
         ("granite", "Granite"),
+        ("gemma3", "Gemma3ForConditionalGeneration"),
+        ("gemma3_text", "Gemma3TextModel"),
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/configuration.py:113]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

`ShieldGemma2Config` 已经加入 `CONFIG_MAPPING_NAMES`，但 `MODEL_NAMES_MAPPING` 没有对应的 `shieldgemma2` 条目。`AutoModel.from_config(ShieldGemma2Config(...))` / `AutoModel.from_pretrained(...)` 在 `_LazyAutoMapping.__contains__` 中会因为该 `model_type` 不在模型映射里返回 `False`，随后抛出无法解析模型类的异常；也就是 PR 说明里的 AutoModel 注册对 ShieldGemma2 实际不可用。请补上模型映射，并增加一个 `AutoModel.from_config(ShieldGemma2Config(...))` 或 `AutoModel.from_pretrained` 的回归测试。

```suggestion
        ("gemma3_text", "Gemma3TextModel"),
        ("shieldgemma2", "ShieldGemma2ForImageClassification"),
```

File: paddleformers/transformers/auto/processing.py [hunk 2]
```diff
@@ -55,6 +55,7 @@
         ("qwen3_omni_moe", "Qwen3OmniMoeProcessor"),
         ("paddleocr_vl", "PaddleOCRVLProcessor"),
         ("ernie4_5_moe_vl", "Ernie4_5_VLProcessor"),
+        ("gemma3", "Gemma3Processor"),
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/processing.py:58]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

当前只注册了 `gemma3 -> Gemma3Processor`，但 ShieldGemma2 标准 checkpoint 会通过 `processor_class=ShieldGemma2Processor` 走 `AutoProcessor.from_pretrained`。本 PR 没有提供 `ShieldGemma2Processor`，`processor_class_from_name("ShieldGemma2Processor")` 会返回 `None`，而此时 `config` 也还没加载到 `PROCESSOR_MAPPING` 分支，最终只能退化成 tokenizer/image processor，丢失 ShieldGemma2 需要的图文预处理、policy/padding 行为。请补上专用 processor（可以复用 Gemma3 的图像处理，但需要显式处理 ShieldGemma2 的 processor_class/策略逻辑）并注册/导出，同时加一个 `AutoProcessor.from_pretrained` 或 processor class 解析测试。

期望修复形态示例：

```python
# paddleformers/transformers/shieldgemma2/processor.py
class ShieldGemma2Processor(Gemma3Processor):
    ...

# paddleformers/transformers/auto/processing.py
("shieldgemma2", "ShieldGemma2Processor")

# paddleformers/transformers/shieldgemma2/__init__.py 和顶层 __init__.py
"processor": ["ShieldGemma2Processor", "ShieldGemma2ProcessorKwargs"]
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查当前提交。之前没有 Nyanpasu 已发布的未解决线程；这次发现两处需要先修复的注册问题，详情已放在行级评论中。

另外，当前检查里 CLA 仍处于 pending，合入前也需要处理。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/shieldgemma2/processor.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    def __call__(
+        self,
+        images: ImageInput | None = None,
+        text=None,
+        **kwargs: Unpack[ShieldGemma2ProcessorKwargs],
+    ) -> BatchFeature:
+        if not images:
+            raise ValueError("ShieldGemma 2 needs images to classify")
+        if not isinstance(images, Sequence):
+            images = [images]
```
Comment (risemeup1111) [inline paddleformers/transformers/shieldgemma2/processor.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

`ImageInput` 在本仓库定义里包含单张 `np.ndarray` 和 `paddle.Tensor`，但这里用 `if not images` 会先对数组/张量做布尔判断；例如 numpy 图像会抛出 truth-value ambiguous 异常，导致 `ShieldGemma2Processor(images=np_image, ...)` 在进入图像预处理前失败。请只用 `None`/空序列判断缺失输入，并保留单图 wrap 逻辑。

```suggestion
        if images is None:
            raise ValueError("ShieldGemma 2 needs images to classify")
        if not isinstance(images, Sequence):
            images = [images]
        elif len(images) == 0:
            raise ValueError("ShieldGemma 2 needs images to classify")
```

File: paddleformers/transformers/auto/processing.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/auto/processing.py:58]: 已复查当前 head：`PROCESSOR_MAPPING_NAMES` 已新增 `("shieldgemma2", "ShieldGemma2Processor")`，并补了 `ShieldGemma2Processor` 导出和 `processor_class_from_name("ShieldGemma2Processor")` 测试。这条原问题已由新提交修复；新的输入兼容性问题我已另在当前 diff 上评论。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交，之前的 AutoModel 映射和 AutoProcessor/ShieldGemma2Processor 注册问题已在当前代码里补齐。新增 processor 实现里还有一个会影响标准 `ImageInput` 的兼容性问题，详细见 inline 评论；该问题需要修复后再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by a31413510** [comment]
代码规范问题
image

**Comment by a31413510** [comment]
自行评论 /re-run all-failed 重跑失败ci
